README
Text-CSV-Easy

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Text::CSV::Easy

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Text-CSV-Easy

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Text-CSV-Easy

    CPAN Ratings
        http://cpanratings.perl.org/d/Text-CSV-Easy

    Search CPAN
        http://search.cpan.org/dist/Text-CSV-Easy/


LICENSE AND COPYRIGHT

Copyright (C) 2013 Tom Peters

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Text-CSV-Easy

0.53    2014-01-22
        Fixes an issue with 0 being treated as undef

0.52    2013-02-09
        POD documentation updates.

0.51    2013-02-09
        Slightly optimizes the regexp for CSV parsing.
        Properly handles CRLF.
        Improved documentation.

0.50    2013-02-09
        Supports RFC 4180.

0.05    2013-02-08
        Update metadata

0.04    2013-02-08
        Fixes an issue when a comman follows an escaped quote.
        Adds TCE_VERSION numbering which will only load XS if TCE_VERSION matches.

0.03    2013-02-07
        Better handles empty strings and undef values. 
        Adds tests to ensure XS and PP compatability.

0.02    2013-02-06
        Minor changes.

0.01    2013-02-06
        First version, released on an unsuspecting world.


lib/Text/CSV/Easy_PP.pm
package Text::CSV::Easy_PP;
use 5.010;
use strict;
use warnings FATAL => 'all';

use Carp;
use Exporter qw(import);

our @EXPORT_OK = qw(csv_build csv_parse);

=head1 NAME

Text::CSV::Easy_PP - Easy CSV parsing and building implemented in PurePerl

=head1 SYNOPSIS

  use Text::CSV::Easy_PP qw(csv_build csv_parse);

  my @fields = csv_parse($string);
  my $string = csv_build(@fields);

=head1 DESCRIPTION

Text::CSV::Easy_PP is a simple module for parsing and building CSV lines. This module is written in PurePerl. For a faster alternative, see L<Text::CSV::Easy_XS>. Either way, you should just be able to use L<Text::CSV::Easy> directly and it will determine which version to use.

This module conforms to RFC 4180 (L<http://tools.ietf.org/html/rfc4180>) for both parsing and building of CSV lines.

=over 4

=item 1. Use commas to separate fields. Spaces will be considered part of the field.

 abc,def, ghi        => ( 'abc', 'def', ' ghi' )

=item 2. You may enclose fields in quotes.

 "abc","def"         => ( 'abc', 'def' )

=item 3. If your field contains a line break, a comma, or a quote, you need to enclose it in quotes. A quote should be escaped with another quote.

 "a,b","a\nb","a""b" => ( 'a,b', "a\nb", 'a"b' )

=item 4. A trailing newline is acceptable (both LF and CRLF).

 abc,def\n           => ( 'abc', 'def' )
 abc,def\r\n         => ( 'abc', 'def' )

=back

When building a string using csv_build, all non-numeric strings will always be enclosed in quotes.

=head1 SUBROUTINES

=head2 csv_build( List @fields ) : Str

Takes a list of fields and will generate a CSV string. This subroutine will raise an exception if any errors occur.

=cut

sub csv_build {
    my @fields = @_;
    return join ',', map {
        if ( !defined )
        {
            '';
        }
        elsif (/^\d+$/) {
            $_;
        }
        else {
            ( my $str = $_ ) =~ s/"/""/g;
            qq{"$str"};
        }
    } @fields;
}

=head2 csv_parse( Str $string ) : List[Str]

Parses a CSV string. Returns a list of fields it found. This subroutine will raise an exception if a string could not be properly parsed.

=cut

sub csv_parse {
    my ($str) = @_;

    return () unless $str;

    my $last_pos = 0;

    my @fields;
    while (
        $str =~ / (?:^|,) 
          (?: ""                # don't want a capture group here
            | "(.*?)(?<![^"]")" # find quote which isn't being escaped
            | ([^",\r\n]*)      # try to match an unquoted field
          )
          (?:\r?\n(?=$)|)       # allow a trailing newline only
          (?=,|$) /xsg
        )
    {
        my $field = $1 || $2;

        # is the field a numeric 0.
        if ( defined($field) && $field =~ /^0+$/ ) {

            # don't do anything.
        }
        else {

            # if we don't have a value, we have either an undef or an empty string.
            # "" will be an empty string, otherwise it should be undef.
            $field ||= ( $& =~ /^,?""(?:\r?\n)?$/ ? "" : undef );
        }

        # track the pos($str) to ensure each field happends immediately after the
        # previous match. also, account for a leading comma when $last_pos != 0
        croak("invalid line: $str")
            if pos($str) > $last_pos + length($&) + ( $last_pos != 0 ? 1 : 0 );

        $last_pos = pos($str);

        if ($field) {
            croak("quote is not properly escaped")
                if ( $field =~ /(?<!")"(?!")/ );

            # unescape the quotes.
            $field =~ s/""/"/g;
        }
        push @fields, $field;
    }

    croak("invalid line: $str") if $last_pos != length($str);

    return @fields;
}

1;

=head1 SEE ALSO

=over 4

=item L<Text::CSV>

=item L<Text::CSV::Easy>

=item L<Text::CSV::Easy_XS>

=back

=head1 AUTHOR

Thomas Peters, E<lt>weters@me.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Thomas Peters

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/Text/CSV/Easy.pm
package Text::CSV::Easy;
use strict;
use warnings FATAL => 'all';
use 5.010;

use B qw(svref_2object);
use Exporter qw(import);

=head1 NAME

Text::CSV::Easy - Easy CSV parsing and building

=head1 VERSION

Version 0.53

=cut

our $VERSION = '0.53';

our @EXPORT_OK = qw(csv_build csv_parse);

# used to ensure XS and PP stay in sync.
our $TCE_VERSION;
BEGIN { $TCE_VERSION = 2 }

=head1 SYNOPSIS

 use Text::CSV::Easy qw( csv_build csv_parse );
 $csv = csv_build(@fields);
 @fields = csv_parse($csv);

=head1 DESCRIPTION

Text::CSV::Easy is a simple module for parsing and building CSV strings. This module itself is
a lightweight wrapper around L<Text::CSV::Easy_XS> or L<Text::CSV::Easy_PP>.

This module conforms to RFC 4180 (L<http://tools.ietf.org/html/rfc4180>) for both parsing and building of CSV lines.

=over 4

=item 1. Use commas to separate fields. Spaces will be considered part of the field.

 abc,def, ghi        => ( 'abc', 'def', ' ghi' )

=item 2. You may enclose fields in quotes.

 "abc","def"         => ( 'abc', 'def' )

=item 3. If your field contains a line break, a comma, or a quote, you need to enclose it in quotes. A quote should be escaped with another quote.

 "a,b","a\nb","a""b" => ( 'a,b', "a\nb", 'a"b' )

=item 4. A trailing newline is acceptable (both LF and CRLF).

 abc,def\n           => ( 'abc', 'def' )
 abc,def\r\n         => ( 'abc', 'def' )

=back

When building a string using csv_build, all non-numeric strings will always be enclosed in quotes.

=cut

my $MODULE;

BEGIN {
    my $xs_loaded = eval { require Text::CSV::Easy_XS; 1 };
    if ( $xs_loaded && $Text::CSV::Easy_XS::TCE_VERSION == $TCE_VERSION ) {
        $MODULE = 'Text::CSV::Easy_XS';
    }
    else {
        $MODULE = 'Text::CSV::Easy_PP';
        require Text::CSV::Easy_PP;
    }

    no strict 'refs';
    for ( keys %{ $MODULE . '::' } ) {

        if ( defined &{"${MODULE}::$_"} ) {
            my $ref = \&{"${MODULE}::$_"};
            my $obj = svref_2object($ref);
            next unless $obj->isa('B::CV') && $obj->GV->STASH->NAME eq $MODULE;
            *$_ = $ref;
        }
    }
}

=head1 SUBROUTINES

=head2 csv_build( List @fields ) : Str

Takes a list of fields and will generate a csv string. This subroutine will raise an exception if any errors occur.

=head2 csv_parse( Str $string ) : List[Str]

Parses a CSV string. Returns a list of fields it found. This subroutine will raise an exception if a string could not be properly parsed.

=head2 module( ) : Str

Returns the underlying module used for CSV processing.

=cut

sub module { return $MODULE }

1;

__END__

=head1 TCE_VERSION

Version 2

The TCE_VERSION ensures that the XS and PP versions of Text::CSV::Easy are consistent.

=head1 SEE ALSO

=over 4

=item L<Text::CSV>

=item L<Text::CSV::Easy_PP>

=item L<Text::CSV::Easy_XS>

=back

=head1 AUTHOR

Thomas Peters, E<lt>weters@me.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Thomas Peters

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=cut
